fedimint-core/src/db/mod.rs

```diff
@@ -354,674 +400,762 @@ impl Database {
+        self.inner.raw_remove_by_prefix(&key).await
     }
+}
 
+#[apply(async_trait_maybe_send!)]
+impl<Inner> IDatabaseTransactionOps for PrefixDatabaseTransaction<Inner>
+where
+    Inner: IDatabaseTransactionOps,
+{
     async fn rollback_tx_to_savepoint(&mut self) -> Result<()> {
-        self.0
-            .as_mut()
-            .context("Cannot rollback to a savepoint on an already consumed transaction")?
-            .rollback_tx_to_savepoint()
-            .await?;
-        Ok(())
+        self.inner.rollback_tx_to_savepoint().await
     }
 
     async fn set_tx_savepoint(&mut self) -> Result<()> {
-        self.0
-            .as_mut()
-            .context("Cannot set a tx savepoint on an already consumed transaction")?
-            .set_tx_savepoint()
-            .await?;
-        Ok(())
+        self.set_tx_savepoint().await
     }
 }
 
+/// Core raw operations database transactions support
 #[apply(async_trait_maybe_send!)]
-impl<'a, Tx: IDatabaseTransaction<'a> + MaybeSend> ISingleUseDatabaseTransaction<'a>
-    for SingleUseDatabaseTransaction<'a, Tx>
+pub trait IDatabaseTransactionOpsCore: MaybeSend {
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: &[u8]) -> Result<Option<Vec<u8>>>;
+
+    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>>;
+
+    async fn raw_remove_entry(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>>;
+
+    /// Returns an stream of key-value pairs with keys that start with
+    /// `key_prefix`. No particular ordering is guaranteed.
+    async fn raw_find_by_prefix(&mut self, key_prefix: &[u8]) -> Result<PrefixStream<'_>>;
+
+    /// Same as [`Self::raw_find_by_prefix`] but the order is descending by key.
+    async fn raw_find_by_prefix_sorted_descending(
+        &mut self,
+        key_prefix: &[u8],
+    ) -> Result<PrefixStream<'_>>;
+
+    /// Delete keys matching prefix
+    async fn raw_remove_by_prefix(&mut self, key_prefix: &[u8]) -> Result<()>;
+}
+
+#[apply(async_trait_maybe_send!)]
+impl<T> IDatabaseTransactionOpsCore for Box<T>
+where
+    T: IDatabaseTransactionOpsCore + ?Sized,
 {
-    async fn commit_tx(&mut self) -> Result<()> {
-        self.0
-            .take()
-            .context("Cannot commit an already committed transaction")?
-            .commit_tx()
-            .await
+    async fn raw_insert_bytes(&mut self, key: &[u8], value: &[u8]) -> Result<Option<Vec<u8>>> {
+        (**self).raw_insert_bytes(key, value).await
     }
 
-    fn add_notification_key(&mut self, key: &[u8]) -> Result<()> {
-        self.0
-            .as_mut()
-            .context("Cannot add notification on an already consumed transaction")?
-            .add_notification_key(key)
+    async fn raw_get_bytes(&mut self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        (**self).raw_get_bytes(key).await
     }
-}
 
-// TODO: use macro again
-#[doc = " A handle to a type-erased database implementation"]
-#[derive(Clone)]
-pub struct CommitTracker {
```

Comment:
Do we still log when there's a write but no commit?

Reply:
Still there. Just moved and diff shows it as removed + added ?